### DISCUSSION TITLE
Need for a practical C# example

### DISCUSSION THREAD

**Discussion by dinther** [mpaperno/WASimCommander#17] [General]
I am trying to write a simple test app using C# WPF where I want to receive data changes for "AUTOPILOT HEADING LOCK DIR" and "HEADING INDICATOR"

I spent almost a week trying to write my test app and I am getting nowhere. I trip up on so many issues that are not answered by the sample app.

1 - How can I have an arbitrary number of requestID's The example only defines an enum at compile time. When I try to pass a uint my wpf form disappears. Are there some threading issues?

2 - What is the difference between a DataRequest and DataRequestRecord. In particular the latter confuses me because the debugger shows 
![image](https://github.com/mpaperno/WASimCommander/assets/1192916/5172fd98-3a3f-47b8-b6b8-2c5f948acf2e) yet I can not access any of those properties. 
3 - How do I reconcile between `client.saveDataRequest(new DataRequest(...`  and what I receive back in the `DataSubscriptionHandler` ?

Wishlist

The C# implementation of WASimCommander should be less like C++ and instead be more oriented towards meeting the needs of developers who want to communicate with MSFS.

I wish the WASim C# wrapper would deal with the weird enum thing for the request identifiers. The trick I pull in my simconnectplus wrapper doesn't seem to work here.  I would also like to see a single object representing a variable. In other words the object I create and pass into saveDataRequest should be the same object that I get back in the DataSubscriptionHandler with an easy to read value property.

I really wanted to write a simple C# WPF sample app manipulating some graphics in WPF using WASimCommander and make it available to be included with WASimCommander SDK but I am unable to do it.

![image](https://github.com/mpaperno/WASimCommander/assets/1192916/7de76588-bb94-4bd2-9ddd-5368264ed88d)

I need help. I have an application to write but I am sorry to say WASimCommander is only adding to my problems at the moment.

**Comment by mpaperno** [comment]
Hi Paul,

> How can I have an arbitrary number of requestID's The example only defines an enum at compile time. When I try to pass a uint my wpf form disappears. Are there some threading issues?

`DataRequest.requestId` _is_ a `uint` type. It's up to your program what value to use here. You can generate them sequentially if you want, just don't re-use them. This is really just like SimConnect works when mapping a data definition ID. Except it _doesn't_ require an `Enum` type as the ID... so it's _less_ restrictive.

I'm sorry I have no idea what this has to do with your form disappearing or what you mean by "threading issues."  Perhaps if I could see some code...?

> What is the difference between a DataRequest and DataRequestRecord.

Sorry, but at risk of repeating what is already written, have you seen the documentation?
<http://wasimcommander.max.paperno.us/struct_w_a_sim_commander_1_1_c_l_i_1_1_structs_1_1_data_request_record.html#details>

>  In particular the latter confuses me because the debugger shows...

I can't tell what it's showing you there... `dr` should be a `DataRequestRecord` type, which is what my VS2022 debugger shows in that spot (but seems like you added a line or maybe other code... 🤷🏼). 

> ... yet I can not access any of those properties.

Access in terms of what?  Which properties?  The console example logs `dr.requestId` and name right there... and all the other properties are also public.

> How do I reconcile between `client.saveDataRequest(new DataRequest(...` and what I receive back in the `DataSubscriptionHandler` ?

Based on the `DataRequestRecord.requestId` which is going to be what _you_ specified in the original `DataRequest`.

To correlate with your "SimConnectPlus," you can then retrieve your `SimVar` object from whatever storage it's stored in based on this ID.... just like you do it now for SimConnect's variable requests (the "definition").

> I wish the WASim C# wrapper would deal with the weird enum thing for the request identifiers. 

If you mean the `DataRequest[Record].requestId` then I think I already explained three's no "enum thing." Otherwise not sure what you mean here.

> In other words the object I create and pass into saveDataRequest should be the same object that I get back in the DataSubscriptionHandler...

I'm not sure what you mean by "same object" here at all... as in the same instance of an object you created and passed to `WASimClient`?  That is a bad idea for a number of reasons, threading issues being one of them, responsibility for holding the original reference being another.

>  ... with an easy to read value property.

I suppose "easy to read" is relative, but that's essentially exactly what you get back in `DataSubscriptionHandler`.  `DataRequestRecord` is a copy of your original `DataRequest` plus two additional properties holding the value data and update timestamp.

Coming back to your "SimConnectPlus" project, I _think_ what you're asking in the last few quotes is to allow you to do away with keeping track of your own `SimVar` objects.  Well, technically you could do that already since `WASimClient` does keep a list of requests (see [`dataRequests()`](http://wasimcommander.max.paperno.us/class_w_a_sim_commander_1_1_client_1_1_w_a_sim_client.html#a6851ac9c59355be4a16b0ca3647bbfa3)/[`C#`](http://wasimcommander.max.paperno.us/class_w_a_sim_commander_1_1_c_l_i_1_1_client_1_1_w_a_sim_client.html#a5c8bc4bc4d480af0d61ea6746b713120), but I would not recommend that (as per the docs, this is returned by copy -- actual objects are not shared for aforementioned reasons).  Besides the actual object storage/sharing issues, you're also better off creating your own "SimVar" object types which are going to contain the information _your program_ needs, not limited (or overburdened) by what "the WASimClient author thinks you _may_ need."

Or maybe I misunderstand completely...  :-)

> Need for a practical C# example

I'm not sure if this is going to help or confuse at this point, but I have what I consider a very practical C# program using `WASimClient` to which I link from the main README: [MSFS Touch Portal Plugin](https://github.com/mpaperno/MSFSTouchPortalPlugin/tree/next/MSFSTouchPortalPlugin). It has no GUI, so I also don't don't if you consider that "practical" and it's certainly too complex overall for an "example."  But among other things, it allows access to any number of sim "variables" (which could be calculated values, not just actual vars) that are completely configurable by the user at runtime. Which I'd say is about as dynamic as it can get.  Same for setting vars and triggering events. It uses both `WASimClient`
and `SimConnect` and decides which is going to be most efficient for a particular purpose, or can fall back to SimConnect-only if `WASimModule` isn't detected at the sim (with more limited functionality, of course).

If you care to peruse, I think you may find the [`SimConnectService`](https://github.com/mpaperno/MSFSTouchPortalPlugin/blob/next/MSFSTouchPortalPlugin/Services/SimConnectService.cs) and [`SimVarItem`](https://github.com/mpaperno/MSFSTouchPortalPlugin/blob/next/MSFSTouchPortalPlugin/Types/SimVarItem.cs) of interest as starting points. The "main" program code (which uses `SimConnectService` and where most `SimVarItem`s are created) is in [`PluginService`](https://github.com/mpaperno/MSFSTouchPortalPlugin/blob/next/MSFSTouchPortalPlugin/Services/PluginService.cs), but beyond that it probably doesn't hold much relevance.

> I really wanted to write a simple C# WPF sample app manipulating some graphics in WPF

Do you have that first part working, manipulating graphics?  Eg. with dummy data?  Before involving WC or SC at all, I mean.

> I need help.

Please share actual code you've tried so far. I think this will be a more efficient way to approach whatever difficulties you're experiencing and perhaps find more practical solutions WC could provide.

Thanks,
-Max

**Reply by dinther** [reply]
Thanks for your patience Max, I fear my frustration leaked through in my post. My apologies for that. I didn't want to reply until I knew what I was talking about. It appears the issue was a threading issue in my WPF program.  The code is bare-bones for readability.

The init section registers two variables with saveDataRequest
``` C#
client.OnDataReceived += DataSubscriptionHandler;

client.saveDataRequest(new DataRequest(
	(uint)0,
	'A',
	"AUTOPILOT HEADING LOCK DIR",
	ValueTypes.DATA_TYPE_FLOAT,
	UpdatePeriod.Millisecond,
	interval: 20,
	deltaEpsilon: 0.1f)
);

client.saveDataRequest(new DataRequest(
	(uint)1,
	'A',
	"HEADING INDICATOR",
	ValueTypes.DATA_TYPE_FLOAT,
	UpdatePeriod.Millisecond,
	interval: 20,
	deltaEpsilon: 0.001f)
);
```

``` C#
private void DataSubscriptionHandler(DataRequestRecord dr)
{
	dr.tryConvert(out float fVal);
	switch (dr.requestId)
	{
       		case 0: headingBugBlock.Text = fVal.ToString(); break;
		case 1: headingBlock.Text = (fVal * 57.2958).ToString(); break;
	}
}
```

The above handler `DataSubscriptionHandler` produces this error:

![image](https://github.com/mpaperno/WASimCommander/assets/1192916/86f8a42b-76a5-4f51-affe-2d7e2bdd39fc)

The wide spread online advice is to wrap a Dispatcher.Invoke around it:
``` C#
private void DataSubscriptionHandler(DataRequestRecord dr)
{
	this.Dispatcher.Invoke(() =>
	{
		dr.tryConvert(out float fVal);
		switch (dr.requestId)
		{
       			case 0: headingBugBlock.Text = fVal.ToString(); break;
			case 1: headingBlock.Text = (fVal * 57.2958).ToString(); break;
		}
	});
}
```

However, that causes the WPF window to disappear. The thing that threw me off was the fact that if I only registered one variable with saveDataRequest the window did show and the value would be rendered correctly.

Eventually I found the answer on this page https://itecnote.com/tecnote/wpf-application-is-deadlocked-when-invoking-on-the-dispatcher/ resulting in this code:

``` C#
private void DataSubscriptionHandler(DataRequestRecord dr)
{
	dr.tryConvert(out float fVal);
	Action action = null;
	switch (dr.requestId)
	{
       		case 0: action = () => { headingBugBlock.Text = fVal.ToString(); }; break;
		case 1: action = () => { headingBlock.Text = (fVal * 57.2958).ToString(); }; break;
	}
	if (action != null) Dispatcher.BeginInvoke(action);
}
```

And this event handler works correctly. It renders both values to the window.
But, just because this works doesn't mean it is correct so I'd welcome any suggestions to do things the right way.

Below the full source code both XAML and .cs

```
<Window x:Name="mainWindow" x:Class="WASimWPF.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:WASimWPF"
        mc:Ignorable="d"
        Title="MainWindow" Height="450" Width="800" Closing="Window_Closing">
    <Grid>
        <StackPanel Margin="10,10,10,10">
            <TextBlock x:Name="headingBugBlock" Text="TextBlock" TextWrapping="Wrap"/>
            <TextBlock x:Name="headingBlock" Text="TextBlock" TextWrapping="Wrap"/>
        </StackPanel>
    </Grid>
</Window>
```

``` C#
using System;
using System.Windows;
using WASimCommander.CLI;
using WASimCommander.CLI.Enums;
using WASimCommander.CLI.Structs;
using WASimCommander.CLI.Client;

namespace WASimWPF
{
    public partial class MainWindow : Window
    {
		WASimClient client;

		public MainWindow()
        {
            InitializeComponent();
			client = new WASimClient(1964);
			client.connectSimulator();
			//UInt32 version = client.pingServer();
			client.connectServer();
			client.OnDataReceived += DataSubscriptionHandler;

			client.saveDataRequest(new DataRequest(
					(uint)0,
					'A',
					"AUTOPILOT HEADING LOCK DIR",
					ValueTypes.DATA_TYPE_FLOAT,
					UpdatePeriod.Millisecond,
					interval: 20,
					deltaEpsilon: 0.1f)
			);

			client.saveDataRequest(new DataRequest(
					(uint)1,
					'A',
					"HEADING INDICATOR",
					ValueTypes.DATA_TYPE_FLOAT,
					UpdatePeriod.Millisecond,
					interval: 20,
					deltaEpsilon: 0.001f)
			);
		}

		private void DataSubscriptionHandler(DataRequestRecord dr)
		{
			dr.tryConvert(out float fVal);
			Action action = null;
			switch (dr.requestId)
			{
				case 0:	action = () => { headingBugBlock.Text = fVal.ToString(); }; break;
				case 1: action = () => { headingBlock.Text = (fVal * 57.2958).ToString(); }; break;
			}
			if (action != null) Dispatcher.BeginInvoke(action);
		}

        private void Window_Closing(object sender, System.ComponentModel.CancelEventArgs e)
        {
			client.disconnectServer();
			client.disconnectSimulator();
			client.Dispose();
		}
	}
}
```

**Reply by mpaperno** [reply]
Hi Paul,

OK yeah, that makes sense... the `DataSubscriptionHandler()` is being called from another thread (started by `WASimClient` in the C++ library) which doesn't own the GUI resources.  And doesn't look like I documented the delivery thread anywhere specifically, so apologies for that.

I really don't know much about working with WPF in particular, but `Dispatcher.Invoke()` or `BeginInvoke()` seem reasonable for this situation, and in fact document this kind of usage specifically (eg. https://learn.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.begininvoke).  

For more background on the issue you may find some of the top answers on this Stack question interesting: https://stackoverflow.com/questions/14703698/invokedelegate

Another which suggests a couple ways to systematically marshal GUI manipulation to the main thread: https://stackoverflow.com/questions/3159580/invoke-a-delegate-on-the-main-thread-in-a-tiered-architecture

Your first version, using `Invoke()`, needs to copy  (capture) `DataRequestRecord` to use inside the invoked function.  If you change the syntax to be like your working version, but simply use `Invoke()` at the end instead of `BeginInvoke()` it should work as well (though using the latter async version seems reasonable in this case anyway). But the main point is that only the actual GUI control manipulation part (`headingBlock.Text = x`) needs to happen on the main GUI thread, everything else should be fine in the thread the delegate was invoked in to begin with.

Besides documenting the multi-threaded delivery in `WASimClient`, I'm not entirely sure what else would be helpful in this case (to ensure event delegates are invoked only in a particular thread).  I guess one way would be to do what SimConnect does, where your code has to call `ReceiveMessage()` explicitly and hence controls the thread in which messages are delivered.  Though this always seemed rather awkward to me.  

I'll think on this some more, and am open to suggestions.

And BTW it would indeed be great to have a GUI example to include!  This looks like a nice start.

A couple minor points... the minimum refresh interval is 25ms (check [`UpdatePeriod.Millisecond`](http://wasimcommander.max.paperno.us/namespace_w_a_sim_commander_1_1_enums.html#a43b33b172f6f4b43947d3a6da9847044) docs). And `deltaEpsilon` is maybe not quite what you expect there... eg. `0.1` means "only send changes which are greater than 0.1", so a change of `10.1 -> 10.2` would not be sent, but `10.1 -> 10.21` would be.  If you want to eliminate all changes which are smaller than one tenth, you'd want to use something like `0.09` for the delta epsilon.

Cheers,
-Max

**Reply by dinther** [reply]
Thank you Max, I will share a public project and a wrapper that meets my desires more. Then you pick as you please for your project.

C# is fairly new to me. Having to provide a time interval from the GUI end is kinda upside down.

Usually it's the GUI asking for updates based on rendering speed. 

EDIT:  Actually, it can be up to the application to disconnect that. Receive and store changes as they happen but a separate loop tied to frame rendering that consumes the stored values.

**Reply by dinther** [reply]
Hi Max, I published what I have so far on my github https://github.com/dinther/SimCom
Have a look see what you think. A wrapper is probably the most appropriate approach but maybe there are ideas that you like for your library.

I also have a question. SimConnect allows me to register and receive SimConnect events. If I am correct these are K variables?
I have not found a way yet to receive these using WASimCommander. Any suggestions where to look for info?

**Reply by mpaperno** [reply]
Hi Paul, you've been busy!

>  SimConnect allows me to register and receive SimConnect events. If I am correct these are K variables?

The "K" vars aren't really variables. They're called that in the RPN docs because you can use calculator code to trigger Key Events by name using the ">K:" syntax/prefix.  But you cannot read/get a "K" value, because they're actually transient events.  "HTML Events" ('H' "vars") are the same way.

What SimConnect allows you to do is subscribe to notifications about when a Key Event was triggered. Which could be either due to an actual device input, UI interaction, or anything else that triggers Key Events (including your own SimConnect client, if you want that).

Personally I have never found a use for that because generally the status of everything one can control with Key Events can be checked with Sim Vars already. And anyway a transient event isn't all that useful for a "remote status" type display because if the event is missed (eg. the UI isn't even running yet) then the UI is going to be out of sync anyway (so, the UI can only use data it can receive on demand).  At any rate, didn't see any reason to recreate that functionality in WC.

**Reply by mpaperno** [reply]
> Usually it's the GUI asking for updates based on rendering speed.

You also bring this up in your readme notes in SimCom...   I'm wondering, when would you use such a paradigm?  I rally can't think of one practical example where the UI refresh/redraw "readiness" state forces a redraw if nothing has changed... why would it redraw anything if no data changes?

(I know there are some kinds of LCD displays and maybe others that do require constant refresh, but this is hardware-level stuff... I'm talking about high level GUI rendering here. Even in low-level embedded programming one doesn't usually copy video memory buffers around unless one has to.)

Typically a redraw is scheduled to run ASAP once there is something new to render.  So for example in your gauge animation I'd transform each part only in response to an actual data update from the sim for that particular part of the gauge... eg. the bug indicator and knob when the bug value changes, the radial indicator/knob when radial or heading changes, etc.  There's no need for the tick timer at all, as far as I can see.  That's the beauty of getting data changes pushed to you... there's nothing to do until data actually changes.

**Reply by dinther** [reply]
Got ya on the even front. When I was trying to configure a full autopilot panel I found sometimes I could not find the corect SimVar's and resorted to events. You make some good points though. I can always implement that part in the same wrapper using simconnect.

A proper way I should use the redraw on the UI side is to first set a flag when data is received and thus changed and in my UI rendering only re-render when that flag is set. In fact each SimVal can have it's own flag and that would do it.

I need to read-up on how c# WPF does the rendering. Something like a requestAnimationFrame in Javascript. Such routine would be best to have it ask WASimCommander to update and then react to the datachange events.

Again something for me to tinker with and see what works best. Mind you, I still need an elegant way to get past the thread issue. I thought a separate timer on the UI side was a good way.

But what do you think of the way I call getVariable calling it with or without an interval value and only dealing with the one SimVal object?

**Reply by dinther** [reply]
Not sure how C# WPF renders the window either. It could be advantageous to set all of the changed values in one go in the case of my HSI instrument example. Since they are all set at once there will only be one render required. Setting different values on each OnDataReceived could be more costly.

I am writing the demo your way now.

**Reply by mpaperno** [reply]
`requestAnimationFrame` is a browser-specific feature to help synchronize animations to the current screen's refresh rate.  It's basically a scheduler for your program, like setInterval(), but optimized.

On the desktop you don't need this... or rather if you need to sync an animation to the monitor refresh rate, there are going to be various other ways to do that (though not really with WPF, perhaps).  But this is really relevant for game-quality animations, to prevent tearing and such, and even then not necessarily how often the scene needs to actually redraw (eg. smooth 30Hz animation on a 60Hz screen).  And yea you wouldn't use WPF if you need that kind of control...  :)

As for the gauge, you just queue your redraws up as you need to... the windowing/render system is going to batch them anyway if possible, and only the actual changed pixels of the UI get updated anyway.  The actual transforms will be basically "free" until the scene is actually rendered.

**Reply by dinther** [reply]
Max, you said that you can do everything with Sim Variables. I can read NAV OBS:1 to find out the Radial the instrument is set to. But it is a read-only SimVar. This is why I ended up mixing events because I can set it using the Event VOR1_SET. Despite being an ex 737 jockey, this Fancy GPS computer screen stuff has me stumped. It seems to have something to do with Nav 1. Steam gauge any day. :-(

Oh I did the demo your way
``` C#
        private void SimCom_OnDataChanged(SimCom SimCom, SimVal SimVal, dynamic NewValue)
        {
            if (SimVal == HeadingBug)
            {
                HeadingBugIndicator.Angle = HeadingBug.Value;
                HeadingBugKnob.Angle = HeadingBug.Value * 5;
            }
            if (SimVal == Heading)
            {
                RadialIndicator.Angle = -Heading.Value + Radial.Value;
                HeadingIndicator.Angle = -Heading.Value;
            }
            if (SimVal == Radial)
            {
                RadialIndicator.Angle = -Heading.Value + Radial.Value;
                RadialKnob.Angle = Radial.Value * 5;
            }
            if (SimVal == Aircraft)
            {
                this.Title = Aircraft.Value;
            }
        }
```

That is what made me worry about extra renders but you are probably right. Still the above code doesn't work unless I use my option 1 and pass the windows Dispatcher to SimCom to do a BeginInvoke.

It's fun and worth it to go through this process. Learning lot's

**Reply by mpaperno** [reply]
The computer stuff can be a quagmire for sure.  And it can vary between the different computer systems, and models... especially 3rd party models where they try to get everything working, usually involving a bunch of L vars and whatnot.  Best bet for 3rd party models is if they have their own docs for this stuff.

But regardless, I didn't say you can do "everything" with Sim Vars... I said you can usually get the status of things with 'em.  _Some_ are settable, and in many cases that's the only way to control something, but there are 1000+ Key Events which also need to be used for other stuff.  :)

So it's common that you use an event to trigger something and a var to check the status of that same thing.

You mean you actually flew 737s in RL? Cool!  Well at least you then know how it's _supposed_ to work... LOL.  Not sure if that ultimately helps or is just more frustrating...  😆 

And yea, you'll still need to dispatch the GUI update parts (transforms) of course.  The dispatcher option you have there in `SimCom` seem like a nice way to handle it.  That's kinda like passing a Window handle to SimConnect constructor, if you think about it (so SC ends up running the message dispatcher in the GUI thread).

**Reply by dinther** [reply]
It's clear by the different ways of naming of the SimVar's and Events alone that it has been growing for a long time. I have developed a range of USB input devices and I am now doing the windows software part. I try to streamline all the variable stuff to make it easier for the users to customize what it does.

Flying 737 is a long time ago. 1994 till 1996 I was stationed in Jakarta flying B737-300 and -400 for Garuda. The -300 had a few screens and a FMC but it was before the GPS days. We had an inertial navigation unit and then of course the old VOR and NDB stuff.
Looking at a cockpit now and I am just about as flabbergasted as any novice. Too many tiny things to distract.

Anyway, I call it quits now. I worked through the night. (New Zealand here)
Thanks for all your help Max. I make sure to send you one of these USB devices when ready.

**Comment by mpaperno** [comment]
A USB device? For simming?  And it needs a GUI?  That's pretty much one of my specialties...  😉 
Anyway, you have me intrigued. And thanks! NZ is on my "overdue to visit" list, so maybe I can save you the postage...  😆

**Reply by dinther** [reply]
Yeah, I used to do more the high level programming languages but I am an electrical engineer by education. I don't like the massive dev stacks we have these days. So I find myself moving back to micro controllers, PCB design and robotics. It is just insane how little this stuff costs now.

I am extremely excited by https://www.infusevr.com/ They beat me to it but I am just one person. Take your input devices with you into the VR space. The MSFS API is totally inadequate for such a task but these guys just use an overlay.

**Comment by dinther** [comment]
I am still spending time on my wrapper for WASimCommander. Today I added a SimConnectEventReceiver to my SimCom wrapper which connects directly to SimConnect. SimCom registers those "variables" of type K with the receiver and when the event occurs the SimCom.OnDataChanged is called in exactly the same way as with Sim Variables.

I managed to get SimConnect to work without a window handle as per your instructions. Thanks for that.

Quite pleased with being able to do this:
``` C#
GearPos = simCom.getVariable("(A:GEAR LEFT POSITION,number) (A:GEAR RIGHT POSITION,number) + (A:GEAR CENTER POSITION,number) +",25, 0.2);
```
And just monitor the one value, (I know. We have `GEAR TOTAL PCT EXTENDED` ). However, I am left with two questions. I hope it is appropriate to ask it here because this thread is getting rather long.

1
From the documentation I understand that executeCalculatorCode only returns a fail when there is a connection issue or Timeout. How do I know that that calculatorCode syntax is correct? From my tests the Result value with bad calculatorCode still returns something. `"(A:GEAR LEFT POSITION,number) poopiedoodle +"` still returns the value of the valid reference. I would like to know when there is a syntax error.

2
executeCalculatorCode seems to require units to be included like `"(A:GEAR LEFT POSITION,number)"` without `,number` it returns 0. However, I can make a dataRequest without units.
``` C#
dataRequest = new DataRequest(
    requestId: definitionIndex,
    resultType: CalcResultType.Double,
    calculatorCode: simVal.FullName,
    valueSize: (uint)WaSim_ValueTypes.FLOAT,
    period: updatePeriod,
    interval: Math.Max(interval, 25),
    deltaEpsilon: 0.0f
);
```
What is the reason for this?

**Reply by mpaperno** [reply]
Hi Paul, sorry didn't see the questions in your edit until now.

> How do I know that that calculatorCode syntax is correct?

You don't, not w/out looking at the MSFS Dev mode console.  I've not found any way.  The Gauge API functions happily return `true` even for complete garbage code. That includes all 3 functions, `gauge_calculator_code_precompile()`, `execute_calculator_code()` and `format_calculator_string()`.  :-(

> still returns the value of the valid reference.

Not sure what you mean here... reference to what?

> I can make a dataRequest without units...
> `calculatorCode: simVal.FullName,`

But you're specifying a code string in this request, not for a named variable. So I'm not sure what you mean about "without units."  I can't tell what the calc string itself is.  When the request is a `Named` type, then you can provide the unit name separately from the var name, and the server does the "right thing" with it.

And yes, a unit name is required in calc strings for all variable types which require units. :)  Which are A, C, E, and they're optional for L (which is another topic).

BTW I'm finishing up a batch of changes, among which are some further "abstractions" of how variables can be requested -- for example strings can now be returned for 'A' var Get commands or data requests, by way of the server automatically translating the request to a calculator expression to get the string result (which is the only way to get strings with the gauge API).
